### ISSUE TITLE
Regression on charset/collate from v3.2.2 to v3.3.2

### ISSUE THREAD

**Issue by marforon** [doctrine/dbal#5321]
### Bug Report

|    Q        |   A
|------------ | ------
| Version     | 3.3.*

#### Summary / Current behaviour / How to reproduce

I'm having issue with schema diff after upgrade from 3.2.2 to 3.3.2 (MySQL v8, not defined default_table_options or charset on dbal url connection string). Historically some of my tables were created with utf8 charset as a default and columns had a charset and collate defined as utf8mb4 somehow:
```sql
CREATE TABLE `notification_event` (
   `name` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8_unicode_ci;
```
After upgrade to 3.3.2 Doctrine Migrations wants to change columns and remove charset and collate on column definition. It was fine before.
```sql
ALTER TABLE notification_event CHANGE name name VARCHAR(255) NOT NULL;
```

#### Expected behaviour

Tables created before with utf8mb3 should see columns with charset and collate as before. Only tables with utf8mb4 shouldn't declare custom charset for columns.

**Comment by morozov** [comment]
@marforon how exactly does one reproduce the issue you're experiencing? The SQL in the description doesn't seem to be enough. What version of MySQL are you using?

FWIW, on MySQL 8, the table created from the DDL in the description is introspected like this:
```
mysql> CREATE TABLE `notification_event` (
    ->    `name` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL
    -> ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8_unicode_ci;
Query OK, 0 rows affected, 2 warnings (0.03 sec)

mysql> SELECT t.TABLE_COLLATION,
    ->        ccsa.CHARACTER_SET_NAME
    -> FROM information_schema.TABLES t
    ->     INNER JOIN information_schema.`COLLATION_CHARACTER_SET_APPLICABILITY` ccsa
    ->         ON ccsa.COLLATION_NAME = t.TABLE_COLLATION
    -> WHERE TABLE_TYPE = 'BASE TABLE' AND TABLE_SCHEMA = 'doctrine' AND TABLE_NAME = 'notification_event';
+-----------------+--------------------+
| TABLE_COLLATION | CHARACTER_SET_NAME |
+-----------------+--------------------+
| utf8_unicode_ci | utf8               |
+-----------------+--------------------+
1 row in set (0.00 sec)
```
Note that the charset name is `utf8`, not `utf8mb3`.

**Comment by marforon** [comment]
@morozov I think it's fine as you have it. Try to do `SHOW CREATE TABLE notification_event; ` and see if the create table is with charset utf8mb3 or not, I just used SHOW CREATE TABLE query to dump the schema definition. But it was created by migration sql as:
```sql
$this->addSql('CREATE TABLE notification_event (name VARCHAR(255) CHARACTER SET utf8mb4 NOT NULL COLLATE `utf8mb4_unicode_ci`) DEFAULT CHARACTER SET utf8 COLLATE `utf8_unicode_ci` ENGINE = InnoDB COMMENT = \'\' ');
```

My exact version is 8.0.26. (Not sure which version it was when I first run this migration, but it doesn't matter in my case, I can replicate it from the scratch on this version)

So to replicate this:

1 Require doctrine/dbal 3.2.2
2. Create an entity with only name property `#[ORM\Column(length: 255)]`
3. Run sql:
```sql
CREATE TABLE `notification_event` (
   `name` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8_unicode_ci;
```
4.  (Clear cache) Check schema diff (for Symfony bin/console doctrine:schema:update --dump-sql) with result
```
[OK] Nothing to update - your database is already in sync with the current entity metadata.
```
5. Require doctrine/dbal 3.3.2
6. (Clear cache) Check schema diff (for Symfony bin/console doctrine:schema:update --dump-sql) with result
```sql
 The following SQL statements will be executed:

     ALTER TABLE notification_event CHANGE name name VARCHAR(255) NOT NULL;
```

**Comment by morozov** [comment]
Could you try reproducing this issue without the ORM and Symfony?

**Comment by marforon** [comment]
@morozov sure, hope it helps:
```php
$connection = DriverManager::getConnection(['url' => 'mysql://root:root@127.0.0.1:3306/test?serverVersion=8.0']);
$connection->executeQuery('CREATE TABLE IF NOT EXISTS `test` (`name` VARCHAR(255) CHARACTER SET utf8mb4 NOT NULL COLLATE `utf8mb4_unicode_ci`) DEFAULT CHARACTER SET utf8 COLLATE `utf8_unicode_ci` ENGINE = InnoDB COMMENT = \'\';');
$schemaManager = $connection->createSchemaManager();
$fromSchema = $schemaManager->createSchema();

$toSchema = new Schema([], [], $schemaManager->createSchemaConfig());
$toSchema->createTable('test')->addColumn(
    'name',
    Types::STRING,
    ['length' => 255, 'notnull' => true, 'platformOptions' => ['version' => false]]
);

$diff = $schemaManager->createComparator()->compareSchemas($fromSchema, $toSchema);
$diffSql = $diff->toSql($connection->getDatabasePlatform());
```

Try to run it with dbal 3.2.2 and then with dbal 3.3.3, see you the difference in `$diffSql`? With 3.2.2 it should be empty and with 3.3.3 it's not.

**Comment by morozov** [comment]
Thanks, @marforon. I can reproduce the issue.

**Comment by morozov** [comment]
So as far as I understand the code,
1. In the database, you create a table with the following defaults: `` CHARACTER SET utf8 COLLATE `utf8_unicode_ci` ``.
2. The table contains a column that overrides the defaults: `` CHARACTER SET utf8mb4 NOT NULL COLLATE `utf8mb4_unicode_ci` ``.
3. In the code, you define a column that doesn't override the defaults:
   ```php
   $toSchema->createTable('test')->addColumn(
       'name',
       Types::STRING,
       ['length' => 255, 'notnull' => true, 'platformOptions' => ['version' => false]]
   );
   ```

So as a result, you get a DDL that removes the overrides from the column in the database:
```sql
ALTER TABLE test CHANGE name name VARCHAR(255) NOT NULL
```

What exactly do you believe is wrong here?

**Comment by surikman** [comment]
@morozov Im not sure, if this is same problem, but previous version of dbal accepts 

```
DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` 
```
on database and during migrations this data was read. After I upgraded to 3.3.3 and run `migrations:diff` all `down` migrations wants to change `COLLATE` on every string (varchar) field, instead of using default value of its table

**Comment by surikman** [comment]
@morozov same test, but little bit changed -> 

```php
$connection = DriverManager::getConnection(['url' => 'mysql://root:root@127.0.0.1:3306/test?serverVersion=8.0']);
$connection->executeQuery('CREATE TABLE IF NOT EXISTS `test` (`name` VARCHAR(255) NOT NULL) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB COMMENT = \'\';');
$schemaManager = $connection->createSchemaManager();
$fromSchema = $schemaManager->createSchema();

$toSchema = new Schema([], [], $schemaManager->createSchemaConfig());
$toSchema->createTable('test')->addColumn(
    'name',
    Types::STRING,
    ['length' => 255, 'notnull' => true, 'platformOptions' => ['version' => false]]
);

$up = $schemaManager->createComparator()->compareSchemas($fromSchema, $toSchema);
$diffSql = $up->toSql($connection->getDatabasePlatform());
assert(count($diffSql) === 0);


$down = $schemaManager->createComparator()->compareSchemas($toSchema, $fromSchema);
$diffSql = $down->toSql($connection->getDatabasePlatform());
assert(count($diffSql) === 0);
```

it generates for `down`:
```sql
ALTER TABLE test CHANGE name name VARCHAR(255) NOT NULL COLLATE `utf8mb4_unicode_ci`
```

but `COLLATE` is also as default for table `test`


I figure out some hint - if I compare schemas without platform (without call `columnsEqual` on it) `up` & `down` are empty

But I upgraded from lower version than 3.2.2 (because schema compare was refactored in #4746 now I am not sure if this is scope of this issue or I need to create a new one)

**Comment by morozov** [comment]
@surikman I believe the problem you're experiencing is different.

The `$up` migration is empty as expected since neither the deployed column nor its DBAL definition override the table defaults. The `$down` one is not. Here's what's happening:

1. Before comparing schemas, the DBAL normalizes them by sourcing the defaults from the "from" schema assuming that this is the currently deployed schema: https://github.com/doctrine/dbal/blob/dfdb5fb2b2230237bf46761140ce9b2dd4edea39/src/Platforms/MySQL/Comparator.php#L34-L37
2. In the "down" case, when calling `compareSchemas()`, the "from" schema is the `$toSchema` which is defined in the DBAL and doesn't have any defaults. I.e. the defaults are empty.
3. The "to" schema (which is `$fromSchema`) is the one that was introspected from the database, so its "test" column contains the options inherited from its table: `{charset = "utf8mb4" collation = "utf8mb4_unicode_ci"}`.
4. In order to migrate from "no options" to "some options", the DBAL generates:
   ```sql
   ALTER TABLE test CHANGE name name VARCHAR(255) CHARACTER SET utf8mb4 NOT NULL COLLATE `utf8mb4_unicode_ci`
   ```

The assumption on step 1 that the "from" is the currently deployed schema is valid for the DBAL (which only migrates the schema forward) but is invalid for Migrations which tries to generate the backward migration by swapping the arguments.

We probably need a separate API that Migrations could use to generate the backward migration.

**Comment by surikman** [comment]
@morozov I create a new issue for this - #5322  (I am not sure if the title is correct, if not feel free to update)

**Comment by surikman** [comment]
@morozov ok I figured out, that this bug is same as in https://github.com/doctrine/migrations/issues/1244

**Comment by greg0ire** [comment]
> We probably need a separate API that Migrations could use to generate the backward migration.

https://github.com/doctrine/dbal/blob/5b6eb6c8ce65ebdc60b0c0960a676cf76758dbf2/src/Schema/Comparator.php#L26

Ok so here we'd need to document that `$fromSchema` is assumed to be the currently deployed schema, and `$toSchema` an in-memory schema / the desired schema.

We should then add another method with a similar signature, so the first step would be to brainstorm an appropriate name for it. Here are my thoughts:
- `compareSchemasBackwards`
- `reverseCompareSchemas`

**Comment by morozov** [comment]
When I worked on the schema-aware comparison, I attempted to rename the arguments from using the "from and to" terminology to something like "actual and desired". But it wasn't straightforward, so I decided not to proceed.

Regarding the naming, I believe it's no longer about the comparison (i.e. the diff), it's about patching in the context of the current connection (i.e. the default SQL Server collation is sourced from the connection). Before brainstorming, we really need to look through the codebase and agree on the semantics of the API (i.e. the exact meaning of the methods and their parameters).
